### DISCUSSION TITLE
How to replace a AcroForm text field by an image

### DISCUSSION THREAD

**Discussion by varandasi** [gettalong/hexapdf#294] [Q&A] [answered]
Hello, 

I'm new to the PDF manipulation and creation. I have a request to set a image field programmatically.

It sounds crazy, but a workaround that popped into my head, was to add a text field with the right position and dimensions, read it and overlap an image.

<img width="586" alt="image" src="https://github.com/gettalong/hexapdf/assets/56271/b8cf2768-1af1-45ba-bd15-3690fa2d431c">

Couldn't find a way, it's more difficult than working with images.

Thanks in advance

**Comment by gettalong** [comment]
The issue title says "text field" and the description talks about an image field, so I'm not sure what you really mean.

Generally, an "image field" in AcroForm doesn't really exist. However, it can be emulated using a button with an icon only layout that uses a Javascript action to update that icon on clicking. Could you provide a sample form that you want to fill out?

If it is not necessary to preserve the form field in question, you could get the positions and dimensions from the field's widget(s), remove the widget(s) and just paint the image at the correct location.

**Reply by varandasi** [reply]
"If it is not necessary to preserve the form field in question, you could get the positions and dimensions from the field's widget(s), remove the widget(s) and just paint the image at the correct location." 

That exactly it! :) I'm trying to workaround not having the possibility to set a acroform image field.

Don't know how to do that, could you point me to some examples? Did searched and read a lot but its not trivial for someone just starting in the pdf world.

Thanks

**Reply by gettalong** [reply, marked as answer]
So, basically, you need to do this:

* Each field can have one or more widgets (see https://hexapdf.gettalong.org/documentation/api/HexaPDF/Type/AcroForm/Field/index.html#method-i-each_widget).
* Each of those widgets is a dictionary and has a `:Rect` entry specifying the rectangle it occupies.
* Iterate over all widgets of the respective field, get the `:Rect` entry, draw the image using the overlay canvas (https://hexapdf.gettalong.org/documentation/api/HexaPDF/Type/Page.html#method-i-canvas) inside that rectangle, and then delete the widget from the field (https://hexapdf.gettalong.org/documentation/api/HexaPDF/Type/AcroForm/Field/index.html#method-i-each_widget)

**Reply by varandasi** [reply]
Thank you for your quick responses, very helpful!

**Reply by varandasi** [reply]
Just sharing this code, might be helpful for someone. There's probably a way to get the page from the form or field
```
image_field = doc.acro_form.field_by_name('image')
image_field.each_widget do |widget|
  rect = widget[:Rect]
  page = doc.pages[1]
  canvas = page.canvas(type: :overlay)
  canvas.image("image.png", at: [rect.left, rect.bottom], width: rect.width)
  image_field.delete_widget(widget)
end
```
